internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -757,3 +786,133 @@ func EnqueueForCompositionRevisionFunc(of resource.CompositeKind, list func(ctx
+// as the conditions arg. All existing conditions will be wiped.
+func ForceSetConditions(u *kunstructured.Unstructured, conditions ...xpv1.Condition) {
+	forceSetConditions(u.Object, "status.conditions", conditions...)
+}
+
+// ForceSetClaimConditions sets the composite's status.claimConditions to whatever is provided
+// as the conditions arg. All existing conditions will be wiped.
+func ForceSetClaimConditions(c *composite.Unstructured, conditions ...xpv1.Condition) {
+	forceSetConditions(c.Object, "status.claimConditions", conditions...)
+}
+
+func forceSetConditions(o map[string]interface{}, path string, conditions ...xpv1.Condition) {
+	conditioned := xpv1.ConditionedStatus{}
+	conditioned.SetConditions(conditions...)
+	_ = fieldpath.Pave(o).SetValue(path, conditioned.Conditions)
+}
+
+// GetConditions returns all items from the object's status.conditions.
+func GetConditions(u *kunstructured.Unstructured) map[xpv1.ConditionType]xpv1.Condition {
+	return getConditions(u.Object, "status.conditions")
+}
+
+// GetClaimConditions returns all items from the composites's status.claimConditions.
+func GetClaimConditions(xr *composite.Unstructured) map[xpv1.ConditionType]xpv1.Condition {
+	return getConditions(xr.Object, "status.claimConditions")
+}
+
+func getConditions(o map[string]interface{}, path string) map[xpv1.ConditionType]xpv1.Condition {
+	conditions := []xpv1.Condition{}
+	if err := fieldpath.Pave(o).GetValueInto(path, &conditions); err != nil {
+		return make(map[xpv1.ConditionType]xpv1.Condition)
+	}
+	m := make(map[xpv1.ConditionType]xpv1.Condition)
+	for _, c := range conditions {
+		m[c.Type] = c
+	}
+	return m
+}
+
+// RemoveStaleConditions takes the previous and current list of conditions.
+// It keeps any special conditions that are required (Ready/Synced)
+// It removes any previous conditions that were not found in the current list.
+// It then merges the remaining previous conditions with the current conditions,
+// giving precedence to previous conditions for sake of accurate transition times.
+func RemoveStaleConditions(prev, curr map[xpv1.ConditionType]xpv1.Condition) []xpv1.Condition {
+	filtered := []xpv1.Condition{}
+	// remove any unknown types that were not seen in the current list
+	for t, p := range prev {
+		if _, ok := curr[t]; ok {
+			filtered = append(filtered, p)
+		} else if t == xpv1.TypeReady || t == xpv1.TypeSynced {
+			filtered = append(filtered, p)
+		}
+	}
+	// add all current entries
+	for _, c := range curr {
+		filtered = append(filtered, c)
+	}
+	return filtered
+}
+
+// TODO(dalton): come up with better naming, break the function up, etc...?
+// if we break this up for sake of better function layout.. it seems we would
+// have to sacrifice performance.
+func handleConditions(rec event.Recorder, o runtime.Object, prev map[xpv1.ConditionType]xpv1.Condition, curr map[xpv1.ConditionType]FunctionCondition) []xpv1.Condition {
```

Comment:
Info: This function removes any conditions that are no longer being returned by the composition functions. Additionally, for any new or updated condition, it records an event. I added this as I did not observe any existing behavior on the Claim or XR that records events when conditions change (only manual calls to `record`).